### PR TITLE
Use HTTP Range response for Playlist player

### PR DESCRIPTION
**Pull request by sangwoo108**
For large videos, we should send data in small pieces. This patch adds supports for HTTP range requests to the `WebUIURLLoaderFactory` and serves the request in the `PlaylistDataSourcd` which inherits `content::URLDataSource`.

With this, we can serve videos without OOM crash.

Resolves https://github.com/brave/brave-browser/issues/34614

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/playlist/playlist_data_source.cc
Comment (fallaciousreasoning) [inline browser/playlist/playlist_data_source.cc:114]: Not sure if this was intentional
```suggestion
    NOTREACHED() << "type is not in {thumbnail,media,favicon}: "
```

File: browser/playlist/playlist_data_source.h
Comment (fallaciousreasoning) [inline browser/playlist/playlist_data_source.h:36]: The docs for `content::WebContents::Getter` say that most usages can be replaced with a `WeakPtr` - is there any reason we use the `Getter` here (and other places in this file) instead?

File: components/playlist/browser/mime_util.cc
Comment (fallaciousreasoning) [inline components/playlist/browser/mime_util.cc:29]: Out of interest, what is it that stops us using net/base/mime_util?

File: components/playlist/browser/mime_util.cc
Comment (fallaciousreasoning) [inline components/playlist/browser/mime_util.cc:119]: nit:
```suggestion
    base::FilePath::StringPieceType file_extension) {
```

File: components/playlist/browser/mime_util.h
Comment (fallaciousreasoning) [inline components/playlist/browser/mime_util.h:14]: might be worth a comment explaining why we can't use the Chromium function

File: components/playlist/browser/mime_util.h
Comment (fallaciousreasoning) [inline components/playlist/browser/mime_util.h:20]: nit:

```suggestion
    base::FilePath::StringPieceType file_extension);
```

**Review by fallaciousreasoning — APPROVED** [review]
LGTM, but I think it'd be good if someone with a bit more mojom expertise could have a look at `ReadRangeData` in `chromium_src/content/browser/webui/web_ui_url_loader_factory.cc`

File: browser/playlist/playlist_data_source.h
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.h:36]: I guess content::WebContents::Getter is just a legacy that should be removed.  The `weakptr` part for the doc was updated later, leaving technical debt.

https://chromium-review.googlesource.com/c/chromium/src/+/5190999

File: components/playlist/browser/mime_util.cc
Comment (sangwoo108) [reply components/playlist/browser/mime_util.cc]: Yeah, good question. I've compared the outputs and found that net::base::mime_util doesn't return correct info for some media types like mp2t and 3gpp. These seem to be supported by the media framework, so made our own version to cover that.

File: components/playlist/browser/mime_util.h
Comment (sangwoo108) [reply components/playlist/browser/mime_util.h:17]: Good call!

**Comment by mkarolin** [comment]
I think you may want to add @iefremov and maybe a security review for this cc: @fmarier

File: chromium_src/content/browser/webui/web_ui_url_loader_factory.cc
Comment (goodov) [inline chromium_src/content/browser/webui/web_ui_url_loader_factory.cc:53]: this function can be removed if with this approach https://gist.github.com/goodov/99a01a03dd8839b2986b5e2d14086797

pro: don't need to copy-paste `ReadData` internals.

File: components/playlist/browser/mime_util.cc
Comment (goodov) [inline components/playlist/browser/mime_util.cc:29]: 1. use consexpr FixedFlatMap for mime->fileext map.
2. use flat_map for fileext->mime map.
3. use map.find() instead of contains+at.

**Comment by sangwoo108** [comment]
> I think you may want to add @iefremov and maybe a security review for this cc: @fmarier

Yup, opened a request here - https://github.com/brave/reviews/issues/1529

File: components/playlist/browser/mime_util.cc
Comment (sangwoo108) [reply components/playlist/browser/mime_util.cc:36]: This might not be exactly same what you meant, but updated to use fixed_flat_map for mime=>ext map. In order to pick up the first mimetype when building ext=>mimemap, leaves the array of pairs.

File: components/playlist/browser/mime_util.cc
Comment (sangwoo108) [inline components/playlist/browser/mime_util.cc:117]: I'm not sure why the chromium-style check complains when using `auto` for this. Could be false alarm?

```
../../brave/components/playlist/browser/mime_util.cc:75:7: error: [chromium-style] auto variable type must not deduce to a raw pointer type.
   75 |   if (auto iter = kMimeToExtensionMap.find(mime_type);
      |       ^~~~
      |       auto*
1 error generated.
```

File: browser/playlist/playlist_data_source.cc
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.cc]: Good call!

File: browser/playlist/playlist_data_source.h
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.h]: Done.

File: browser/playlist/playlist_data_source.h
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.h]: Done.

File: browser/playlist/playlist_data_source.cc
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.cc]: Yup, done.

File: browser/playlist/playlist_data_source.cc
Comment (sangwoo108) [reply browser/playlist/playlist_data_source.cc]: Awesome! Changed.

File: chromium_src/content/browser/webui/web_ui_url_loader_factory.cc
Comment (sangwoo108) [reply chromium_src/content/browser/webui/web_ui_url_loader_factory.cc]: A way better! Thanks.

**Review by goodov — APPROVED** [review]
lgtm with few nits.

I've looked into how it's implemented upstream and can confirm that this solution is correct, I haven't found anything else we can do to simplify things further.

Although there's another way of doing that via memmapped files, but:
1. existing upstream implementation uses uint32_t along with `base::checked_cast` to calculate sizes and offsets which will break on files > 4GB on both x86 and x64.
2. big memmapped files won't work on x86 just by the nature of the limited address space.

File: components/playlist/browser/mime_util.cc
Comment (sangwoo108) [reply components/playlist/browser/mime_util.cc:36]: > ah, you're right, I haven't figure out the order that should be preserved to pick the right mimetype (I should've read the comment carefully).

Nah, I updated the comment in the latest commit. 

> Another take: maybe it's easier to just create two fixed flat maps with a clear usecase for each and have a unit test that makes sure both maps are in sync?

Yeah, what I was interested in was keeping maps in sync, so that makes sense. Plus, we can construct both maps at compile time 👍

**Comment by fmarier** [comment]
> maybe a security review for this

@bcaller reviewed as part of 
https://github.com/brave/reviews/issues/1529#issuecomment-1975775666

**Comment by fmarier** [comment]
This might be a good one to look at upstreaming.

File: chromium_src/content/browser/webui/web_ui_url_loader_factory.cc
Comment (sangwoo108) [reply chromium_src/content/browser/webui/web_ui_url_loader_factory.cc]: Good call. Done

**Pull request merged**
